### PR TITLE
story(ccls-1806) correspondence address screen and searching

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**
Includes CCLS-1807

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/bean/AddressFormData.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.laa.ccms.caab.bean;
+
+import lombok.Data;
+
+/**
+ * Represents an address details form.
+ */
+@Data
+public class AddressFormData {
+  private String preferredAddress;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/AddressFormData.java:10]: javadocs for attributes?

File: src/main/java/uk/gov/laa/ccms/caab/bean/AddressSearchFormData.java [hunk 2]
```diff
@@ -6,6 +6,6 @@
  * Represents the client address search form.
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/AddressSearchFormData.java:6]: Change javadoc now that class has been renamed?

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/client/FindAddressValidator.java [hunk 3]
```diff
@@ -33,25 +35,43 @@ public boolean supports(Class<?> clazz) {
...
+
+      country = addressDetails.getCountry();
+      postcode = addressDetails.getPostcode();
+      houseNameNumber = addressDetails.getHouseNameNumber();
+    }
+    validateRequiredField("country", country,
         "Country", errors);
-    if (StringUtils.hasText(addressDetails.getCountry())) {
-      if (!addressDetails.getCountry().equals("GBR")) {
+    if (StringUtils.hasText(country)) {
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/client/FindAddressValidator.java:64]: `if (COUNTRY_CODE_GBR.equals(country))`

(then there's no need to do nested if)

File: src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClientErrorHandler.java [hunk 4]
```diff
@@ -63,6 +64,18 @@ public Mono<ApplicationProviderDetails> handleGetProviderDetailsError(Throwable
...
+   *
+   * @param e the encountered error
+   * @return a Mono signaling the error wrapped in a {@code CaabApiServiceException}
+   */
+  public Mono<Address> handleGetCorrespondenceAddressError(Throwable e) {
+    final String msg = "Failed to retrieve correspondence address";
+    log.error(msg, e);
+    return Mono.error(new CaabApiClientException(msg, e));
+  }
+
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/CaabApiClientErrorHandler.java:78]: renamed

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/client/FindAddressValidator.java (hunk 3 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/client/FindAddressValidator.java:64]: sorted

File: src/main/java/uk/gov/laa/ccms/caab/bean/AddressFormData.java (hunk 1 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/AddressFormData.java:10]: sorted

File: src/main/java/uk/gov/laa/ccms/caab/bean/AddressSearchFormData.java (hunk 2 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/AddressSearchFormData.java:6]: renamed

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 5]
```diff
@@ -411,20 +416,43 @@ public ApplicationFormData getProviderDetailsFormData(final String id) {
...
+   */
+  public void updateCorrespondenceAddress(
+      final String id,
+      final AddressFormData addressFormData,
+      final UserDetail user) {
+
+    final Address correspondenceAddress = addressFormDataMapper.toAddress(addressFormData);
+
+    caabApiClient.putApplication(
+        id, user.getLoginId(), correspondenceAddress, "correspondence-address").block();
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:439]: Constants for address type?

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 6]
```diff
@@ -437,7 +465,7 @@ public void patchApplicationType(
             applicationFormData.getDevolvedPowersContractFlag())
         .build();
 
-    caabApiClient.patchApplication(
+    caabApiClient.putApplication(
         id, user.getLoginId(), applicationType, "application-type").block();
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:469]: constant

**Review by porritta — COMMENTED** [review]
Review complete

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 6 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: amended

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 5 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: amended

**Review by porritta — APPROVED** [review]
LGTM

**Pull request merged**